mesonbuild/interpreter/compiler.py

```diff
@@ -743,3 +747,68 @@ def get_supported_function_attributes_method(self, args: T.Tuple[T.List[str]], k
     @noKwargs
     def get_argument_syntax_method(self, args: T.List['TYPE_var'], kwargs: 'TYPE_kwargs') -> str:
         return self.compiler.get_argument_syntax()
+
+    @typed_kwargs('compiler.symbol_visibility_header',
+        KwargInfo('api', (str, NoneType)),
+        KwargInfo('compilation', (str, NoneType)),
+        KwargInfo('static_compilation', (str, NoneType)),
+        KwargInfo('static_only', bool, default=False),
+        )
+    @typed_pos_args('compiler.symbol_visibility_header', str)
+    @FeatureNew('compiler.symbol_visibility_header', '0.63.0')
+    def symbol_visibility_header_method(self, args: T.List[str], kwargs: 'TYPE_kwargs') -> mesonlib.File:
+        if self.compiler.language not in {'c', 'cpp'}:
+            raise InterpreterException('compiler.symbol_visibility_header is supported only for C and C++ compilers')
+
+        header_name = args[0]
+
+        default_prefix = mesonlib.underscorify(self.interpreter.active_projectname.upper())
+        api = kwargs['api'] or f'{default_prefix}_API'
+        compilation = kwargs['compilation'] or f'{default_prefix}_COMPILATION'
+        static_compilation = kwargs['static_compilation'] or f'{default_prefix}_STATIC_COMPILATION'
+        static_only = kwargs['static_only']
+
+        content = ''
+
+        if static_only:
+            content += textwrap.dedent(f'''\
+                #ifndef {static_compilation}
+                #  define {static_compilation}
+                #endif /* {static_compilation} */
+                ''')
+
+        if self.compiler.info.is_windows():
+            content += textwrap.dedent(f'''\
+                #ifndef {api}
+                #  ifdef {static_compilation}
+                #    define {api} extern
+                #  else /* !{static_compilation} */
+                #    ifdef {compilation}
+                #      define {api} extern __declspec(dllexport)
+                #    else /* !{compilation} */
+                #      define {api} extern __declspec(dllimport)
+                #    endif /* !{compilation} */
+                #  endif /* !{static_compilation} */
+                #endif /* {api} */
+                ''')
+        elif self._has_argument_impl('-fvisibility=hidden'):
+            content += textwrap.dedent(f'''\
+                #ifndef {api}
+                #  ifdef {compilation}
+                #    define {api} __attribute__((visibility("default")))
+                #  else /* !{compilation} */
+                #    define {api}
+                #  endif /* !{compilation} */
+                #endif /* {api} */
```

Comment:
This is incorrect, because if the compiler supports visibility=hidden then there is *never* any circumstance, period, end of story, in which it's bad or wrong to define it.

You do not need to leave it empty when using it as a public header for linking against. You SHOULD NOT leave it empty, because it is confusing to change cases around. Also, it perpetuates the erroneous belief that you need to do so, which means Meson is teaching bad behavior. :)

Reply:
That's entirely subjective to judge if it's bad or not. You don't *need* to define the attribute, but you *can* define it if you want. Both are just equivalently good IMHO.

I copied this from how glib does it, but I don't mind changing it to always define the attribute if it's confirmed that it has no impact.

Reply:
No, it's objectively bad for exactly the social reasons I explained.

Reply:
"social reasons" that's by definition subjective :D

Reply:
Bah, changed to always set the attribute, that's what cmake does too.